src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class CSharpDeterministicKeyBuilder : DeterministicKeyBuilder
+    {
+        internal static readonly CSharpDeterministicKeyBuilder Instance = new();
+
+        private CSharpDeterministicKeyBuilder()
+        {
+        }
+
+        protected override void WriteCompilationOptionsCore(JsonWriter writer, CompilationOptions options)
+        {
+            if (options is not CSharpCompilationOptions csharpOptions)
+            {
+                throw new ArgumentException(null, nameof(options));
+            }
+
+            base.WriteCompilationOptionsCore(writer, options);
+
+            writer.Write("unsafe", csharpOptions.AllowUnsafe);
+            writer.Write("topLevelBinderFlags", csharpOptions.TopLevelBinderFlags);
+            writer.WriteKey("usings");
+            writer.WriteArrayStart();
+            foreach (var name in csharpOptions.Usings)
+            {
+                writer.Write(name);
+            }
+            writer.WriteArrayEnd();
+        }
+
+        protected override void WriteParseOptionsCore(JsonWriter writer, ParseOptions parseOptions)
```

Comment:
Should parseOptions be marked as nullable? (I see that the VB override handles null)

Reply:
It is a bit of a tricky case. The only case where you get `null` is when grabbing the `ParseOptions` from `VisualBasicCompilation`. That only happens when you construct it via the API, it will always have one for command line compilations.

Decided to keep this non-null and add an explicit check in the VB code to guard against it.

Reply:
That's fair. 
nit: Consider adding assertion or comment in C# side